core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/ContainerTorrentDownloadJob.kt

```diff
@@ -0,0 +1,218 @@
+                            println("downloaded torrent in ${System.currentTimeMillis() - startTime} ms")
+                            torrentDeferred.complete(true)
+                        }
+
+                        override fun onProgress(progressSize: Int) {
+                            contentJobItem.cjiItemProgress = (progressSize * 0.75).toLong()
+                            progress.onProgress(contentJobItem)
+                        }
+                    }
+                    GlobalScope.launch(Dispatchers.Default) {
+                        while (true) {
+                            delay(100)
+                            if (!isActive) {
+                                torrentDeferred.completeExceptionally(CancellationException())
+                            }
+                        }
+                    }
+
+                    ustadTorrentManager.addDownloadListener(containerUid, torrentListener as TorrentDownloadListener)
+                    torrentDeferred.await()
+                    ustadTorrentManager.removeDownloadListener(containerUid)
+
+                }
+
+
+                val existingMd5s = existingFiles.mapNotNull { it.cefMd5 }.toSet()
+                val entriesThatGotDownloaded = fileNameList
+                        .filter { it !in existingMd5s || it.equals(MANIFEST_FILE_NAME) }
+
+                val newContainerEntryFiles = mutableListOf<ContainerEntryFile>()
+                entriesThatGotDownloaded.forEach {
+                    val entryFile = ContainerEntryFile().apply {
+                        cefMd5 = it
+                        val cefFile = File(containerFilesFolder, it)
+                        cefPath = cefFile.path
+                        ceTotalSize = cefFile.getUnCompressedSize()
+                        ceCompressedSize = cefFile.length()
+                        compression = if (cefFile.isGzipped()) COMPRESSION_GZIP else COMPRESSION_NONE
+                    }
+                    newContainerEntryFiles.add(entryFile)
+                }
+                db.containerEntryFileDao.insertList(newContainerEntryFiles)
+
+                val manifestFile = File(containerFilesFolder, MANIFEST_FILE_NAME)
+                if (!manifestFile.exists()) throw IllegalArgumentException("no manifest file found")
+
+                val manifestJson = manifestFile.readText()
+                val manifest: ContainerManifest = Json.decodeFromString(
+                        ContainerManifest.serializer(), manifestJson)
+                val newContainerEntry = mutableListOf<ContainerEntry>()
+                manifest.entryMap?.entries?.forEach { entry ->
+                    if (entry.key == MANIFEST_FILE_NAME) {
+                        return@forEach
+                    }
+                    entry.value.forEach { path ->
+                        val cefUid = db.containerEntryFileDao.findEntryByMd5Sum(entry.key.base64EncodedToHexString())?.cefUid
```

Comment:
Any idea how long this function takes? Maybe make one query that would select just cefUid and md5sums needed into one map.